Make create repository dialog behave consistently when initializing a repository

jcansdale opened the pull request:
_See discussion on https://github.com/jcansdale/desktop/pull/1_

## Description

Make the create repository dialog behave in a consistent way when creating a repository or initializing a repository.

- Make the combination of `Name` + `Local Path` point at the dir to be created/initialized
- Disable the `Name` text box when initializing a repository at a read-only path

Fixes https://github.com/desktop/desktop/issues/9651

### Screenshots

#### After

- The `Name` text box is disabled
- The combination of `Name` + `Local Path` points to the directory to be initialized
- The `Description` text box is selected

![image](https://user-images.githubusercontent.com/11719160/80916133-c9c02600-8d4e-11ea-8eb0-3a09e25b04a3.png)

#### Before

- The `Name` text box is enabled (but its contents aren't used)
- The `Name` text box is selected
- `Name` and `Local Path` have the same directory name

![image](https://user-images.githubusercontent.com/11719160/80916098-94b3d380-8d4e-11ea-8bd9-83bac8416d0e.png)

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes:

jcansdale committed:
Disable text box Name when path is read only

jcansdale committed:
Set path to parent dir of initialPath

tierninho reviewed (dismissed):
Took this for a spin, all looks to be functioning correctly in OSX, based on the steps outlined above. ✅  

_(Windows testing is giving me unrelated errors.)_

tierninho commented:
Well, spoke to soon. This just popped up in the console when I clicked on the repo I added:

```
git-process.js:87 Uncaught (in promise) Error: Unable to find path to repository on disk.
    at git-process.js:87
    at exithandler (child_process.js:315)
    at ChildProcess.errorhandler (child_process.js:327)
    at ChildProcess.emit (events.js:203)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:270)
    at onErrorNT (internal/child_process.js:456)
    at processTicksAndRejections (internal/process/task_queues.js:77)
```

jcansdale committed:
Compare initialPath to fullPath

jcansdale commented:
@tierninho,

> Well, spoke to soon. This just popped up in the console when I clicked on the repo I added

That's strange, I haven't been able to repro this. 😕 

I don't know what [this](https://github.com/desktop/desktop/pull/10081/files#diff-650645ed7ae4515fcc9129bb8ee7cb38L193) check is for exactly, but I noticed it was comparing `initialPath` to `path` (which is the parent dir). I've changed it (see https://github.com/desktop/desktop/pull/10081/commits/ccb034f821802b96b403620b016c8efa7c84b9d1) to compare with the combined `fullPath` instead.

I don't _think_ this is related, but I wanted to fix it anyway. 🙂

tierninho commented:
Thanks for the fix. Unfortunately, I still see the error, but perhaps one of the team devs can shed some light on it and we can ignore it. 🤔   I will hold off until the review is done.

jcansdale committed:
Merge remote-tracking branch 'upstream/development' into fixes/9651-git-init-dialog

steveward commented:
Gentle bump since this was requested in https://github.com/desktop/desktop/issues/10774.

niik commented on app/src/ui/add-repository/create-repository.tsx:
```diff
@@ -194,17 +194,21 @@ export class CreateRepository extends React.Component<
+    const fullPath = Path.join(
+      this.state.path,
+      sanitizedRepositoryName(this.state.name)
+    )
+
+    if (this.props.initialPath && this.props.initialPath === fullPath) {
       // if the user provided an initial path and didn't change it, we should
       // validate it is an existing path and use that for the repository
       try {
-        await FSE.ensureDir(currentPath)
-        return currentPath
+        await FSE.ensureDir(fullPath)
+        return fullPath
       } catch {}
     }
 
-    return Path.join(currentPath, sanitizedRepositoryName(this.state.name))
+    return fullPath
   }
```
This method seems odd to me now. If I'm reading this correctly then in the case of `initialPath` being set we'd compare it against our joined components (path and name) which we had previously derived from `initialPath` in the constructor. The only reason they would be different then is if the base name of `initialPath` contained forbidden repo name characters (i.e. `sanitizedRepositoryName(Path.basename(this.props.initialPath)) !== Path.basename(this.props.initialPath`).

If that is the case we would attempt to create the `fullPath`, return `fullPath` if successful, and return `fullPath` if not successful. And the first thing we'd do in `createRepository` which is the only consumer of `resolveRepositoryRoot` is to attempt to create the returned path again.

The result of all of this is that if you have a directory that you want to init at say `~/GitHub/my repo name with spaces` and you select this directory in the Add Existing Repository dialog, and click the `create repository suggestion` you won't actually end up initializing a new repository in `~/GitHub/my repo name with spaces`, instead you'll be creating a brand new folder and initializing it in `~/GitHub/my-repo-name-with-spaces`

So for an initial refactor I believe this entire method could be removed and its callsite replaced with something like this:

```ts
const fullPath =
      this.props.initialPath ??
      Path.join(this.state.path, sanitizedRepositoryName(this.state.name))
```

That doesn't solve the fact that the name textbox will show `my-repo-name-with-spaces` but that's a lesser concern that I don't think we need to or should solve in this PR. Looking at this code I don't even know why we do all the repo name sanitization stuff at all, it's enforcing GitHub's repository naming rules on local repositories which is fine to do in the publish step but not here.

cyberangel67 commented:
I don't agree, the name of the repository should have nothing to do with the path, after all its a name, not a path. And this stops new repositories from being created if a project already exists.

As it stands, if I have a project that I want to create into a repository, I can't do this.

As this is a 6-month-old ticket, why has this issue not been addressed, why am I still not able to take an existing source project and create a repository from that folder? Why does it insist on creating a folder inside that path with the projects name and add nothing to the repository?

In my opinion

Name : should be what you want to name the repsoitory
Description : Self explanatory
local path : the FULL path to the project, and if nothing exists then a path is created.

So it should be

Name: Test Repository Try out
Description : Teting out the creation of a repository.
Local Path: C\test\source\proejcts\TestCode

The name of the repository should not be added to the Local Path! This just confuses the user on how it works, when every other application doesn't work the way this ticket is asking.

tidy-dev committed:
Merge branch 'development' into pr/10081

skyline1983skyliny commented:
> _Zobacz dyskusję na [temat jcansdale#1](https://github.com/jcansdale/desktop/pull/1)_
> 
> ## Opis
> Upewnij się, że okno dialogowe tworzenia repozytorium zachowuje się w spójny sposób podczas tworzenia repozytorium lub inicjowania repozytorium.
> 
> * Tworzenie kombinacji + punkt na dir, który ma zostać utworzony/zainicjowany`Name``Local Path`
> * Wyłączanie pola tekstowego podczas inicjowania repozytorium w ścieżce tylko do odczytu`Name`
> 
> Poprawki #9651
> 
> ### Zrzuty ekranu
> #### Po
> * Pole tekstowe jest wyłączone`Name`
> * Kombinacja + wskazuje katalog, który ma zostać zainicjowany`Name``Local Path`
> * Pole tekstowe jest zaznaczone`Description`
> 
> ![image](https://user-images.githubusercontent.com/11719160/80916133-c9c02600-8d4e-11ea-8eb0-3a09e25b04a3.png)
> 
> #### Przed
> * Pole tekstowe jest włączone (ale jego zawartość nie jest używana)`Name`
> * Pole tekstowe jest zaznaczone`Name`
> * `Name` i mają tę samą nazwę katalogu`Local Path`
> 
> ![image](https://user-images.githubusercontent.com/11719160/80916098-94b3d380-8d4e-11ea-8bd9-83bac8416d0e.png)
> 
> ## Wersji
> Notatki:

dupa

sultannn1234 reviewed (changes requested):
fffffffffffhhhhbhh

niik commented:
Hey @jcansdale, this has been open for a very long time without much activity so I'm going to close this out to keep our PR list manageable but if this is still something you're interested in working on please let us know. There's a comment that I don't even remember writing that would need to be addressed https://github.com/desktop/desktop/pull/10081#discussion_r510867185

jcansdale commented:
Hey @niik,

I'm sorry I haven't had the cycles to play with GitHub Desktop recently. 😞

I do think #9651 is confusing, with what looks like a simple fix. I'd be happy for someone else to take a stab at it.
